gate-web/src/main/groovy/com/netflix/spinnaker/gate/services/CredentialsService.groovy

```diff
@@ -29,21 +28,33 @@ class CredentialsService {
   @Autowired
   ClouddriverService clouddriverService
 
-  List<ClouddriverService.Account> getAccounts() {
+  Collection<String> getAccountNames() {
+    getAccountNames([])
+  }
+
+  Collection<String> getAccountNames(Collection<String> userRoles) {
+    getAccounts(userRoles)*.name
+  }
+
+  Collection<ClouddriverService.Account> getAccounts() {
+    getAccounts([])
+  }
+
+  /**
+   * Returns all account names that a user with the specified list of userRoles has access to.
+   */
+  List<ClouddriverService.Account> getAccounts(Collection<String> userRoles) {
     HystrixFactory.newListCommand(GROUP, "getAccounts") {
-      def allAccounts = clouddriverService.accounts
-
-      if (!AuthenticatedRequest.getSpinnakerUser().present ||
-          !AuthenticatedRequest.getSpinnakerAccounts().present) {
-        // if the request is unauthenticated, return only anonymously accessible accounts (no group membership required)
-        return allAccounts.findAll { !it.requiredGroupMembership }
-      }
-
-      def allowedAccountsOptional = AuthenticatedRequest.getSpinnakerAccounts()
-      def allowedAccounts = allowedAccountsOptional.orElse("").split(",").collect { it.toLowerCase() }
-      return allAccounts.findAll {
-        allowedAccounts.contains(it.name.toLowerCase())
-      }
+      return clouddriverService.accounts.findAll { ClouddriverService.Account account ->
+        if (!account.requiredGroupMembership) {
+          return true // anonymous account.
+        }
+
+        def userRolesLower = userRoles*.toLowerCase()
```

Comment:
The user roles may be possible, but the accounts from clouddriver obviously come from across the wire, and are loaded by Spring's `ConfigurationProperties`, which I don't think has the automatic ability to mutate the values.

That being said, we could override the getter in CD, but we'd have to override the getter for every provider, so between all of the changes that would need to be made, these two lines seem like the path of least resistance.